src/sinks/appsignal/normalizer.rs

```diff
@@ -0,0 +1,199 @@
+use vector_core::event::{Metric, MetricValue};
+
+use crate::sinks::util::buffer::metrics::{MetricNormalize, MetricSet};
+
+#[derive(Default)]
+pub(crate) struct AppsignalMetricsNormalizer;
+
+impl MetricNormalize for AppsignalMetricsNormalizer {
+    fn normalize(&mut self, state: &mut MetricSet, metric: Metric) -> Option<Metric> {
+        // We only care about making sure that counters are incremental, and that gauges are
+        // always absolute. Other metric types are currently unsupported.
+        match &metric.value() {
+            // We always send counters as incremental and gauges as absolute. Realistically, any
+            // system sending an incremental gauge update is kind of doing it wrong, but alas.
+            MetricValue::Counter { .. } => state.make_incremental(metric),
+            MetricValue::Gauge { .. } => state.make_absolute(metric),
+            // Otherwise, send it through as-is.
+            _ => Some(metric),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::BTreeSet;
+
+    use vector_core::event::{Metric, MetricKind, MetricValue};
+
+    use super::AppsignalMetricsNormalizer;
+    use crate::sinks::util::buffer::metrics::{MetricNormalize, MetricSet};
+
+    fn get_counter(value: f64, kind: MetricKind) -> Metric {
+        Metric::new("counter", kind, MetricValue::Counter { value })
+    }
+
+    fn get_gauge(value: f64, kind: MetricKind) -> Metric {
+        Metric::new("gauge", kind, MetricValue::Gauge { value })
+    }
+
+    fn run_comparisons(inputs: Vec<Metric>, expected_outputs: Vec<Option<Metric>>) {
+        let mut metric_set = MetricSet::default();
+        let mut normalizer = AppsignalMetricsNormalizer;
+
+        for (input, expected) in inputs.into_iter().zip(expected_outputs) {
+            let result = normalizer.normalize(&mut metric_set, input);
+            assert_eq!(result, expected);
+        }
+    }
```

Comment:
Well that is lame. somehow GH cut off the rest of my comment.
Basically said it'd be nice to keep the individual test cases, so all the shared unit test functions become `assert_absolute_counter<N: <MetricNormalize>(normalizer: N)` , and in the dd metrics and appsignal tests, you just call that helper function passing in the specific normalizer.
Lmk if needs more clarification.

Reply:
Done!

I deviated a little from your suggested naming. I renamed `run_comparisons` to `assert_normalize`, which seems like a better fit (it asserts the output of calling `.normalize` on a normalizer with certain inputs)

I also gave the test functions more verbose names, describing both the input and the output, since different normalizers can expect different outputs for the same metric type and kind. The names don't start with `assert`, since they're not actually assertions, but full test case templates parameterised on a normalizer, so I put them in a `tests` module within `test_util::metrics`.

I refactored the metric normalizers that were already under thorough testing to use these helpers and test case templates (StatsD, DataDog and AppSignal) but not the other metric normalizers, which are either untested or tested in a different manner. Those can be brought under test more easily with these changes in a different PR/issue.

Let me know if you're happy with the changes!

Reply:
Awesome, yeah my suggestion was meant as a starting point; I appreciate the additional thought and effort you did, thanks!